### PR TITLE
git_config: allow empty values

### LINKED ISSUE #12502: git_config: allow empty string in `user.email`
### Summary

In Git, you can use `git config --global user.email ""` without issue, but in `git_config.py` you cannot.

With `value: ""` or `value: null`, an error is reported:

> "Task failed: Module failed: If state=present, a value must be specified. Use the community.general.git_config_info module to read a config value."

If `null` is provided, the value defaults to an empty string (`""`), which is what I want; however, since Python treats empty strings as falsy, the error condition is triggered and the error is reported.

https://github.com/ansible-collections/community.general/blob/b345875acf6022846e959d7e70c73619db3a7ce7/plugins/modules/git_config.py#L170-L177

Something like the following would fix it:

```python3
# /usr/lib/python3/dist-packages/ansible_collections/community/general/plugins/modules/git_config.py
name_allows_empty_string = { 'user.email', 'aaa', 'bbb' }

# "present" with no "value" and not in "exceptions"
if not unset and not new_value and not name in name_allows_empty_string:
    module.fail_json(
        msg="If state=present, a value must be specified. Use the community.general.git_config_info module to read a config value."
    )
```

I am not sure whether other values for the `name` parameter would behave the same way (Git is immense). I don’t know either whether this is too specific to include, so I wanted to ask the maintainers if this is worth a PR.

### Issue Type

Bug Report

### Component Name

git_config

### Ansible Version

```
ansible [core 2.19.4]
  config file = None
  configured module search path = ['$HOME/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3/dist-packages/ansible
  ansible collection location = $HOME/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/bin/ansible
  python version = 3.13.5 (main, Jul 15 2026, 20:25:40) [GCC 14.2.0] (/usr/bin/python3)
  jinja version = 3.1.6
  pyyaml version = 6.0.2 (with libyaml v0.2.5)
```

### Community.general Version

```console
# /usr/lib/python3/dist-packages/ansible_collections
Collection        Version
----------------- -------
community.general 11.2.1
```

### Configuration

_No response_

### OS / Environment

 Debian GNU/Linux 13 (trixie)

### Steps to Reproduce

```yaml
- name: Set global default git email
  community.general.git_config:
    add_mode: "replace-all"
    name: "user.email"
    value: ""
    # value: null
    scope: "global"
    state: "present"
```

### Expected Results

Being able to set email to an empty string.

### Actual Results

> "Task failed: Module failed: If state=present, a value must be specified. Use the community.general.git_config_info module to read a config value."

### Code of Conduct

- [x] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by ricardbejarano**
##### SUMMARY

Fixes https://github.com/ansible-collections/community.general/issues/12502

`git config user.email ""` is valid, but `git_config` refused it: `value` was normalized with `params["value"] or ""` and then checked with `if not unset and not new_value`, so an empty string was indistinguishable from a missing value and the task failed with `If state=present, a value must be specified.`

As suggested by @felixfontein in the issue, the check is now `new_value is None`. `value: null` therefore still fails (the message is the intended one, and `required_if` already covers the case where the option is omitted entirely), while `value: ""` is accepted.

There was a second empty-string bug behind that one. The current value was read as `out.rstrip().splitlines()`, and since `git config --get-all` prints an empty line for an empty value, `rstrip()` reduced it to `""` and `splitlines()` to `[]` — exactly what an unset setting looks like. A setting with an empty value would therefore never compare equal to the requested empty value and the module would report a change on every run. The output is now split on the value separator instead, dropping only the single trailing newline that `git config --get-all` always emits.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME

git_config

##### ADDITIONAL INFORMATION

The `git_config` integration target is not `unsupported`, so the new `set_empty_value.yml` tasks run in CI: they set an empty value, set it again to assert idempotency, read it back through `git_config_info` (`config_values == {"user.email": [""]}`, which distinguishes an empty value from an unset one, unlike `config_value`), overwrite it with a non-empty value, and finally unset a setting that holds an empty value.

The parsing change is behavior-preserving for non-empty values: `git config --get-all` terminates each value with a newline, so the only difference against `rstrip().splitlines()` is that trailing whitespace of the last value is no longer silently dropped.

### CODE REVIEW

File: changelogs/fragments/12502-git-config-empty-value.yml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+bugfixes:
+  - 'git_config - allow to set a setting to an empty string, the way ``git config user.email ""`` does. Passing an
+    empty ``value`` failed with ``If state=present, a value must be specified``, and once a setting had an empty
+    value the module was no longer idempotent for it
+    (https://github.com/ansible-collections/community.general/issues/12502).'
```
Comment (russoz) [inline changelogs/fragments/12502-git-config-empty-value.yml:5]: The second sentence is overkill for a changelog, please remove it.

And you must include the link to this PR as well.

File: tests/integration/targets/git_config/tasks/set_empty_value.yml [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    value: foo
+    scope: global
+  register: set_result3
+
+- name: assert the empty value was set idempotently
+  ansible.builtin.assert:
+    that:
+      - set_result1 is changed
+      - set_result2 is not changed
+      - 'get_result.config_values == {"user.email": [""]}'
```
Comment (russoz) [inline tests/integration/targets/git_config/tasks/set_empty_value.yml:40]: Why the value of the property is `[""]` and not `""`? That seems to divert from the purpose of the PR and the issue.

Probably better not to depend on git config being empty prior to this. I would replace this with:
```suggestion
      - get_result.config_values["user.email"] == ""
```

**Review by russoz — COMMENTED** [review]
Hi @ricardbejarano 

A couple of adjustments are needed.

Please note that, when using AI, you must abide by the [Ansible Community Policy for AI-Assisted Contributions](https://docs.ansible.com/projects/ansible/latest/community/ai_policy.html#ansible-community-policy-for-ai-assisted-contributions). Thanks!

File: tests/integration/targets/git_config/tasks/set_empty_value.yml (hunk 2 above)
Comment (Kaniee) [reply tests/integration/targets/git_config/tasks/set_empty_value.yml:40]: I think it's because the docs of [git_config_info](https://docs.ansible.com/projects/ansible/latest/collections/community/general/git_config_info_module.html#return-config_values) state, that the return value `config_values` is mapping a setting to a list of values.

So either compare `get_result.config_values["user.email"] == [""]` or use `config_value` which only returns the first element.

File: tests/integration/targets/git_config/tasks/set_empty_value.yml [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    name: user.name
+    state: absent
+    scope: global
+  register: unset_result
+
+- name: assert the empty value was unset
+  ansible.builtin.assert:
+    that:
+      - unset_result is changed
+...
```
Comment (Kaniee) [inline tests/integration/targets/git_config/tasks/set_empty_value.yml:61]: I propose to add a another test to cover the "add" mode with empty strings.
E.g. `name: push.pushOption`. First add `""`, then add `"foo"` and finally compare to `["", "foo"]`.

That way you can test it for multi-valued, with `""` being a reset value.

Note: I've never contributed to this project before, so I don't know if this complies with the testing standards in this repo.

File: changelogs/fragments/12502-git-config-empty-value.yml (hunk 1 above)
Comment (felixfontein) [reply changelogs/fragments/12502-git-config-empty-value.yml:5]: Also the PR's URL is missing.

File: plugins/modules/git_config.py [hunk 4]
```diff
@@ -197,7 +198,11 @@ def main():
         # If the return code is 1, it just means the option hasn't been set yet, which is fine.
         module.fail_json(rc=rc, msg=err, cmd=" ".join(list_args))
 
-    old_values = out.rstrip().splitlines()
+    # 'git config --get-all' terminates every value with a newline, so splitting always yields a
+    # trailing empty element which must not be mistaken for a configured empty value.
+    old_values = out.split("\n")
+    if old_values and old_values[-1] == "":
+        old_values.pop()
```
Comment (felixfontein) [inline plugins/modules/git_config.py:205]: I'm not sure, but aren't you simply reimplementing `old_values = out.splitlines()`?

(Also for `out.split("\n")`, `old_values` is never empty, so the `if` can be simplified to `if old_values[01] == ""`. But simply writing `old_values = out.splitlines()` is even shorter.)

File: tests/integration/targets/git_config/tasks/set_empty_value.yml (hunk 2 above)
Comment (felixfontein) [reply tests/integration/targets/git_config/tasks/set_empty_value.yml:40]: Yep, `get_result.config_values["user.email"] == [""]` sounds like the way to go.

File: tests/integration/targets/git_config/tasks/set_empty_value.yml (hunk 3 above)
Comment (felixfontein) [reply tests/integration/targets/git_config/tasks/set_empty_value.yml:61]: Such a test would be great!

File: tests/integration/targets/git_config/tasks/set_empty_value.yml (hunk 2 above)
Comment (russoz) [reply tests/integration/targets/git_config/tasks/set_empty_value.yml:40]: Why not `config_value` (singular)?

Comment (felixfontein) [reply tests/integration/targets/git_config/tasks/set_empty_value.yml:40]: `config_value` is only working well for options where there's a single value, not a list. (If there are multiple, it simply contains the first one.) To verify something, better use `config_values`.

Comment (russoz) [reply tests/integration/targets/git_config/tasks/set_empty_value.yml:40]: Isn't that exactly the use case addressed by this? One single value `““`

Comment (felixfontein) [reply tests/integration/targets/git_config/tasks/set_empty_value.yml:40]: No. We don't want to check that there is *at least* one value, and the first one is `""`. We want to make sure that there is *only* one value, and it is `""`.

Comment (russoz) [reply tests/integration/targets/git_config/tasks/set_empty_value.yml:40]: Ok, fair enough. IMHO we should ensure that by removing all values by the beginning of the test, but checking for the list also works.

Comment (Kaniee) [reply tests/integration/targets/git_config/tasks/set_empty_value.yml:40]: The additional benefit of checking the list is that we can verify that there is exactly one value, which is `""`.

File: changelogs/fragments/12502-git-config-empty-value.yml [hunk 5]
```diff
@@ -0,0 +1,5 @@
+bugfixes:
+  - 'git_config - allow to set a setting to an empty string, the way ``git config user.email ""`` does. Passing an
+    empty ``value`` failed with ``If state=present, a value must be specified``, and once a setting had an empty
+    value the module was no longer idempotent for it
```
Comment (russoz) [inline changelogs/fragments/12502-git-config-empty-value.yml]: This is bit too verbose for the changelog, IMHO. I would stick with "what changed" and trim the why and the how. The link back to this PR should be enough to explain things, if anyone cares to look for it. 

The outcome of what you write will be published in the collection changelog:

https://docs.ansible.com/ansible/latest/collections/community/general/changelog.html

**Comment by felixfontein** [comment]
Ping @ricardbejarano

needs_info

File: changelogs/fragments/12502-git-config-empty-value.yml (hunk 1 above)
Comment (bejaratommy) [reply changelogs/fragments/12502-git-config-empty-value.yml]: Shortened the changelog and added links to both the issue and this PR in 242d325.

File: plugins/modules/git_config.py (hunk 4 above)
Comment (bejaratommy) [reply plugins/modules/git_config.py]: Indeed. Simplified this to `out.splitlines()` in 242d325.

File: tests/integration/targets/git_config/tasks/set_empty_value.yml (hunk 2 above)
Comment (bejaratommy) [reply tests/integration/targets/git_config/tasks/set_empty_value.yml]: Kept the `config_values` assertion: the return value is list-valued, `setup_no_value.yml` removes the existing config first, and comparing the full mapping verifies there is exactly one value and that it is empty.

File: tests/integration/targets/git_config/tasks/set_empty_value.yml (hunk 3 above)
Comment (bejaratommy) [reply tests/integration/targets/git_config/tasks/set_empty_value.yml]: Added this `add_mode` case in 242d325. It also exposed and fixed `git_config_info` stripping leading empty values from multi-valued output.

**Comment by bejaratommy** [comment]
Addressed all review feedback in `242d325`: shortened the changelog, simplified parsing with `splitlines()`, and added the empty-value `add_mode` integration test. The test exposed and fixed a related `git_config_info` parsing bug.

AI assistance is disclosed with the `Assisted-by: GPT-5` commit trailer.

`validate-modules` and the `git_config` integration target pass locally.

ready_for_review

File: plugins/modules/git_config.py [hunk 6]
```diff
@@ -197,7 +198,7 @@ def main():
         # If the return code is 1, it just means the option hasn't been set yet, which is fine.
         module.fail_json(rc=rc, msg=err, cmd=" ".join(list_args))
 
-    old_values = out.rstrip().splitlines()
+    old_values = out.splitlines()
```
Comment (guenhter) [inline plugins/modules/git_config.py]: Why was striping removed?

File: plugins/modules/git_config_info.py [hunk 7]
```diff
@@ -140,7 +140,7 @@ def main():
         # If the return code is 1, it just means the option hasn't been set yet, which is fine.
         module.fail_json(rc=rc, msg=err, cmd=" ".join(args))
 
-    output_lines = out.strip("\0").split("\0") if out else []
+    output_lines = out[:-1].split("\0") if out else []
```
Comment (guenhter) [inline plugins/modules/git_config_info.py]: Is this the same? If there is no `\0` at the end, the old version would handle this where the new one would just remove the last char, wouldn't it? Is there a reason changing this?

**Review by guenhter — CHANGES_REQUESTED** [review]

File: plugins/modules/git_config.py (hunk 6 above)
Comment (bejaratommy) [reply plugins/modules/git_config.py]: `rstrip()` removed significant whitespace from values, including the empty value this PR adds support for. `splitlines()` removes only the line delimiters while preserving empty values.

File: plugins/modules/git_config_info.py (hunk 7 above)
Comment (bejaratommy) [reply plugins/modules/git_config_info.py]: Good point. Git normally terminates `--null` output with `\0`, but blindly slicing is unnecessarily fragile. Updated in eab1f55 to remove the final NUL only when present.

File: plugins/modules/git_config.py (hunk 6 above)
Comment (guenhter) [reply plugins/modules/git_config.py]: Understood. Thx

**Review by guenhter — APPROVED** [review]

File: tests/integration/targets/git_config/tasks/set_empty_value.yml (hunk 2 above)
Comment (Kaniee) [reply tests/integration/targets/git_config/tasks/set_empty_value.yml]: Didn't we settle on
`get_result.config_values["user.email"] == [""]`?

File: tests/integration/targets/git_config/tasks/set_empty_value.yml [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+- name: getting added values
+  community.general.git_config_info:
+    name: push.pushoption
+    scope: global
+  register: get_added_values
+
+- name: assert empty values work with add mode
+  ansible.builtin.assert:
+    that:
+      - 'get_added_values.config_values == {"push.pushoption": ["", "foo"]}'
```
Comment (Kaniee) [inline tests/integration/targets/git_config/tasks/set_empty_value.yml]: Similar to the other assertion, I'd propose to only assert on the git config that was modified in the test.
```suggestion
      - get_added_values.config_values["push.pushoption"] == ["", "foo"]
```

Comment (russoz) [reply tests/integration/targets/git_config/tasks/set_empty_value.yml]: L63 and L70 are adding these values

File: plugins/modules/git_config_info.py [hunk 9]
```diff
@@ -140,7 +140,7 @@ def main():
         # If the return code is 1, it just means the option hasn't been set yet, which is fine.
         module.fail_json(rc=rc, msg=err, cmd=" ".join(args))
 
-    output_lines = out.strip("\0").split("\0") if out else []
+    output_lines = (out[:-1] if out.endswith("\0") else out).split("\0") if out else []
```
Comment (Kaniee) [inline plugins/modules/git_config_info.py]: **Nit:** I find the two inline `if/else` expressions a bit hard to read.

If the intent is simply to treat a trailing `\0` as a delimiter and remove it, couldn't we simplify this to:

```suggestion
    output_lines = out.removesuffix("\0").split("\0") if out else []
```

This should be functionally equivalent while being easier to read.

If we can rule out `None` and non-empty values without a trailing `\0` as possible values for `out`, then we can further simplify to
```suggestion
    output_lines = out.split("\0")[:-1]
```

File: tests/integration/targets/git_config/tasks/set_empty_value.yml (hunk 8 above)
Comment (Kaniee) [reply tests/integration/targets/git_config/tasks/set_empty_value.yml]: I agree. I propose to only assert on the values of the config `push.pushoption` instead of asserting on the whole git config.
